stubs/regex/regex/regex.pyi

```diff
@@ -0,0 +1,349 @@
+from typing import Any, AnyStr, Callable, NoReturn, Tuple, _Alias, overload
+
+import regex._regex as _regex
+
+__version__: str
+
+A: int
+ASCII: int
+B: int
+BESTMATCH: int
+D: int
+DEBUG: int
+E: int
+ENHANCEMATCH: int
+F: int
+FULLCASE: int
+I: int
+IGNORECASE: int
+L: int
+LOCALE: int
+M: int
+MULTILINE: int
+P: int
+POSIX: int
+R: int
+REVERSE: int
+T: int
+TEMPLATE: int
+S: int
+DOTALL: int
+U: int
+UNICODE: int
+V0: int
+VERSION0: int
+V1: int
+VERSION1: int
+W: int
+WORD: int
+X: int
+VERBOSE: int
+
+DEFAULT_VERSION: int
+
+@overload
+def compile(pattern: AnyStr, flags: int = ..., ignore_unused: bool = ..., **kwargs: Any) -> _regex.Pattern: ...
+@overload
+def compile(pattern: _regex.Pattern, flags: int = ..., ignore_unused: bool = ..., **kwargs: Any) -> _regex.Pattern: ...
```

Comment:
Think most of these methods are right, the main question is around how to handle these custom `Pattern`/`Match`/`Splitter`/`Scanner` objects/types?

Reply:
Because string patterns and bytes patterns behave differently, `Pattern` should probably be generic, so the first overload would say `-> _regex.Pattern[AnyStr]`, specifying which kind of pattern it returns. See `re.Pattern` in stdlib, which is defined in `typing.pyi` for historical reasons.

If `compile()` is supposed to return the pattern as is, you can use a TypeVar:

```python3
_T = TypeVar(bound=_regex.Pattern)

...

@overload
def compile(pattern: _T, flags: int = ..., ignore_unused: bool = ..., **kwargs: Any) -> _T: ...
```

Reply:
@Akuli Thanks, I've made the `Pattern`/`Match`/`Splitter`/`Scanner` generic in `_regex.pyi` 👍 I think it looks pretty consistent with the `re` hinting, but lmk if you think I should also be using the `TypeVar`s as well.